src/analysis/convex/basic.lean

```diff
@@ -117,12 +118,120 @@ end
 lemma segment_translate_preimage (a b c : E) : (λ x, a + x) ⁻¹' [a + b, a + c] = [b, c] :=
 set.ext $ λ x, mem_segment_translate a
 
-lemma segment_translate_image (a b c: E) : (λx, a + x) '' [b, c] = [a + b, a + c] :=
+lemma segment_translate_image (a b c : E) : (λx, a + x) '' [b, c] = [a + b, a + c] :=
 segment_translate_preimage a b c ▸ image_preimage_eq _ $ add_left_surjective a
 
 lemma segment_image (f : E →ₗ[ℝ] F) (a b : E) : f '' [a, b] = [f a, f b] :=
 set.ext (λ x, by simp [segment_eq_image])
 
+/-- Open segment in a vector space. Note that `open_segment x x = {x}` instead of being `∅`. -/
+def open_segment (x y : E) : set E :=
+{z : E | ∃ (a b : ℝ) (ha : 0 < a) (hb : 0 < b) (hab : a + b = 1), a • x + b • y = z}
+
+lemma open_segment_subset_segment (x y : E) :
+  open_segment x y ⊆ [x, y] :=
+λ z ⟨a, b, ha, hb, hab, hz⟩, ⟨a, b, ha.le, hb.le, hab, hz⟩
+
+lemma mem_open_segment_of_ne_left_right {x y z : E} (hx : x ≠ z) (hy : y ≠ z) (hz : z ∈ [x, y]) :
+  z ∈ open_segment x y :=
+begin
+  obtain ⟨a, b, ha, hb, hab, hz⟩ := hz,
+  by_cases ha' : a = 0,
+  { rw [ha', zero_add] at hab,
+    rw [ha', hab, zero_smul, one_smul, zero_add] at hz,
+    exfalso,
+    exact hy hz },
+  by_cases hb' : b = 0,
+  { rw [hb', add_zero] at hab,
+    rw [hb', hab, zero_smul, one_smul, add_zero] at hz,
+    exfalso,
+    exact hx hz },
+  exact ⟨a, b, ha.lt_of_ne (ne.symm ha'), hb.lt_of_ne (ne.symm hb'), hab, hz⟩,
```

Comment:
```suggestion
  by_cases ha' : a ≠ 0,
  by_cases hb' : b ≠ 0,
  { exact ⟨a, b, ha.lt_of_ne (ne.symm ha'), hb.lt_of_ne (ne.symm hb'), hab, hz⟩ },
  all_goals { simp * at * },
```
There are other ways to style this, but as you can see it can be much shorter.

Reply:
But isn't worse compilation-wise? I assume simp takes much longer than specified rewrites, right?

Reply:
I suppose, and you make a fair point, but I think you can deal with it by squeezing the `simp`:
```
 all_goals { simp only [*, add_zero, not_not, one_smul, zero_smul, zero_add] at * },
```
Tbh, I don't know if we prioritize making code as inexpensive as possible...that was the approach I originally took but have since changed my style to emulate others who (seem to) focus more on maximizing readability. If someone who is actually qualified to discuss this (not me by any stretch!) would like to opine I would be much obliged!

Reply:
Okay, I haven't done the change for now. But others feel free to add argument to the discussion.

Reply:
Okay. If you use `simp only` I can't imagine it would be such more intensive (if at all), but there are people here who actually know so listen to them, not me :)